homeassistant/components/toon/oauth2.py

```diff
@@ -0,0 +1,137 @@
+"""OAuth2 implementations for Toon."""
+import logging
+from typing import Any, Optional, cast
+
+from yarl import URL
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import config_entry_oauth2_flow
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from . import config_flow
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def register_oauth2_implementations(
+    hass: HomeAssistant, client_id: str, client_secret: str
+) -> None:
+    """Register Toon OAuth2 implementations."""
+    config_flow.ToonFlowHandler.async_register_implementation(
+        hass,
+        ToonLocalOAuth2Implementation(
+            hass,
+            client_id=client_id,
+            client_secret=client_secret,
+            name="Eneco Toon",
+            tenant_id="eneco",
+            issuer="identity.toon.eu",
+        ),
+    )
+    config_flow.ToonFlowHandler.async_register_implementation(
+        hass,
+        ToonLocalOAuth2Implementation(
+            hass,
+            client_id=client_id,
+            client_secret=client_secret,
+            name="Engie Electrabel Boxx",
+            tenant_id="electrabel",
+        ),
+    )
+    config_flow.ToonFlowHandler.async_register_implementation(
+        hass,
+        ToonLocalOAuth2Implementation(
+            hass,
+            client_id=client_id,
+            client_secret=client_secret,
+            name="Viesgo",
+            tenant_id="viesgo",
+        ),
+    )
+
+
+class ToonLocalOAuth2Implementation(config_entry_oauth2_flow.LocalOAuth2Implementation):
+    """Local OAuth2 implementation for Toon."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        client_id: str,
+        client_secret: str,
+        name: str,
+        tenant_id: str,
+        issuer: Optional[str] = None,
+    ):
+        """Local Toon Oauth Implementation."""
+        self._name = name
+        self.tenant_id = tenant_id
+        self.issuer = issuer
+
+        super().__init__(
+            hass=hass,
+            domain=tenant_id,
+            client_id=client_id,
+            client_secret=client_secret,
+            authorize_url="https://api.toon.eu/authorize",
+            token_url="https://api.toon.eu/token",
+        )
+
+    @property
+    def name(self) -> str:
+        """Name of the implementation."""
+        return f"{self._name} via Configuration.yaml"
+
+    async def async_generate_authorize_url(self, flow_id: str) -> str:
+        """Generate a url for the user to authorize."""
+        data = {"tenant_id": self.tenant_id}
+
+        if self.issuer is not None:
+            data["issuer"] = self.issuer
+
+        url = await super().async_generate_authorize_url(flow_id)
+        return str(URL(url).update_query(data))
```

Comment:
You should use `self.extra_authorize_data()` to add data to the authorize url.

Reply:
`extra_authorize_data` is not part of `LocalOAuth2Implementation` (which this class inherits from. `self.extra_authorize_data` is part of `AbstractOAuth2FlowHandler` in the config flow.

I've picked this method for the following reasoning:

- It keeps all OAuth weirdness for Toon in one place
- It removes the need for me to pass around / fetching the chosen implementation in the config flow to provide that extra authorize data, that is available already in this local implementation.

Reply:
I think that it's a bug that `LocalOAuth2Implementation` doesn't support `extra_authorize_data`. We should add that support.

Reply:
Agreed, fixed, and adjusted this PR.